src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,282 @@
+            //     ])
+            }
+            ItemEnum(ref def, _) => {
+            //     let mut distance_from_explicit = 0;
+            //     (AdtKind::Enum, def.variants.iter().map(|v| {
+            //         let did = tcx.hir.local_def_id(v.node.data.id());
+            //         let discr = if let Some(e) = v.node.disr_expr {
+            //             distance_from_explicit = 0;
+            //             ty::VariantDiscr::Explicit(tcx.hir.local_def_id(e.node_id))
+            //         } else {
+            //             ty::VariantDiscr::Relative(distance_from_explicit)
+            //         };
+            //         distance_from_explicit += 1;
+
+            //         convert_struct_variant(tcx, did, v.node.name, discr, &v.node.data)
+            //     }).collect())
+            }
+
+            ItemStruct(ref def, _) => {
+                // // Use separate constructor id for unit/tuple structs and reuse did for braced structs.
+                // let ctor_id = if !def.is_struct() {
+                //     Some(tcx.hir.local_def_id(def.id()))
+                // } else {
+                //     None
+                // };
+                //let fields = def.fields().iter().map(|f| {
+                //    //let fid = tcx.hir.local_def_id(f.id);
+                //    ////let dup_span = seen_fields.get(&f.name).cloned();
+                //    ////if let Some(prev_span) = dup_span {
+                //    ////    struct_span_err!(tcx.sess, f.span, E0124,
+                //    ////                     "field `{}` is already declared",
+                //    ////                     f.name)
+                //    ////        .span_label(f.span, "field already declared")
+                //    ////        .span_label(prev_span, format!("`{}` first declared here", f.name))
+                //    ////        .emit();
+                //    ////} else {
+                //    ////    seen_fields.insert(f.name, f.span);
+                //    ////}
+
+                //    ty::FieldDef {
+                //        // did: fid,
+                //        did: tcx.hir.local_def_id(f.id),
+                //        name: f.name,
+                //        vis: ty::Visibility::from_hir(&f.vis, node_id, tcx)
+                //    }
+                //}).collect();
+
+
+                // ----------- example
+                for field in def.fields().iter() {
+                    //        for variant in adt_def {
+                    //            for field in variant {
+                    //                // from ty/outlives.rs
+                    //                //   Foo<'b, 'c>  ==> ['b, 'c]
+                    //                //   Vec<T>: 'a
+                    //                //   outlives_components(Vec<T>) = [T]
+                    let outlives = self.tcx.outlives_components(field);
+                    required_predicates.extend(required_predicates_for_type_to_be_wf(field_ty));
```

Comment:
@nikomatsakis stuck on how to call the function `outlives_components`. It requires a `Ty` but not sure how to get that from `field`.

Also could you provide a general idea on where I should look for the implementation of `required_predicates_for_type_to_be_wf`?